Fix issue 10777 - multiSort should return a SortedRange

wilzbach opened the pull request:
As I now bumped for the second time into the problem of having a chainable `multiSort` (=SortedRange), I quickly wrote my own solution. Afterwards I realized that #4061 is already open, there are a couple of minor differences, here
1) a inline, static function instead of MultiPred struct is used.
2) no `shrink` method is provided (but can be in the future)

So in summary less (possibly controversial) changes.

As always, I am very happy about feedback.

@SzMK - let's join efforts to get this into Phobos?

9il commented on std/algorithm/sorting.d:
```diff
@@ -869,31 +873,54 @@ template multiSort(less...) //if (less.length > 1)
+        else
         {
-            while (r.length > 1)
+            void multiSortImpl(funs...)(Range r) @safe
```
Should this be a static fun?

9il commented on std/algorithm/sorting.d:
```diff
@@ -869,31 +873,54 @@ template multiSort(less...) //if (less.length > 1)
+                foreach (f; AliasSeq!(funs))
+                {
+                    alias lessFun = binaryFun!(f);
+                    int r = lessFun(a, b);
```
This will break code, there is no warranty  that lessFun returns `int`. Sometime it may return floating point, long or user defined type.
So the best way is
```d
if (auto r = lessFun(a, b))
    return r;
```

wilzbach commented:
thanks for your input @9il  - addressed ;-)

wilzbach commented on std/algorithm/sorting.d:
```diff
@@ -844,7 +844,7 @@ private template validPredicates(E, less...)
 
 /**
-$(D void multiSort(Range)(Range r)
+$(D auto multiSort(Range)(Range r)
```
is there a way how we could write `SortedRange` here?
The `less` fun is defined inline.

9il replied:
no

wilzbach commented:
Another example of the new Travis linting [at work](https://travis-ci.org/dlang/phobos/builds/126587195) :)

There was also a small issue with autotester, that should be fixed now.

> "What would be other use cases for MultiPred" (from #4061)

Please not that the difference to the stalled #4061 is that we just use a static, pred function which is passed to `assumeSorted`, so this is hidden from the user and if wanted we can still extend add a MultiPred struct in the future if wanted (?).

That being said 1) I would like to see MultiPred too and 2) this should be ready to go. The change for the user is that instead of void a SortedRange will be returned. If so, it will be added to the changelog once #4228 is there.

Ping @MartinNowak.

MartinNowak commented:
Same [comment](https://github.com/dlang/phobos/pull/4061#issuecomment-194098177) as on #4061 applies, converting a multi-sort predicate to a sort predicate should be seen as an intermediate step towards a SortedRange w/ multi-sort support, b/c it doesn't make use of the performance benefits (e.g. lowerBound makes M * log(N) comparisions for M predicates).

MartinNowak commented on std/algorithm/sorting.d:
```diff
@@ -856,12 +856,16 @@ and sorts elements that have the same $(D id) by $(D date)
 < b.id : a.date > b.date"(r)), but $(D multiSort) is faster because it
 does fewer comparisons (in addition to being more convenient).
+
+Returns: The initial range wrapped as a $(D SortedRange) with its predicates.
```
How about `with its predicates converted to an equivalent single predicate.`?

MartinNowak commented on std/algorithm/sorting.d:
```diff
@@ -872,31 +876,53 @@ template multiSort(less...) //if (less.length > 1)
 
-        static if (funs.length > 1)
+        static if (funs.length == 1)
+            return sort!fun[0](r);
```
I wouldn't move the conversion of the predicate to the instantiation of sort w/o a good reason. So please keep the lessFun alias and we're sure to not trigger any weird instantiation error w/ nested templates or run into semantic order issues.

wilzbach replied:
but don't we call `binaryFun` twice then?
(once in `sort` and once here)

wilzbach replied:
I digged a bit deeper and it turns out that
1) this line was __never__ called in Phobos before (see the typo as a proof) -> I added a test to proof the fix!
2) multiSort is not recursive, it doesn't call itself like other templates do - this is just a specialization to sort which will be a bit faster. Hence I don't see any reason why we need to call `binaryFun` twice - we pass in just the alias in a lot of other functions.

MartinNowak commented on std/algorithm/sorting.d:
```diff
@@ -872,31 +876,53 @@ template multiSort(less...) //if (less.length > 1)
-        {
-            sort!(lessFun, ss)(r);
+
+            static auto predFun(ElementType!Range a, ElementType!Range b)
```
no need for return type inference, it's a bool.

MartinNowak commented:
Changelog entry missing.

wilzbach commented:
> Changelog entry missing.

Added (could you please have a look at #4228)?

> b/c it doesn't make use of the performance benefits (e.g. lowerBound makes M * log(N) comparisions for M predicates).

Have you looked at the implementation of `predFun`? It will break as soon as either `lessFun(a,b)` or `lessFun(b, a)` is true. Is there really a better way to do this?

All other comments were addressed - and I added a small fix (`multiSort` didn't work for a single predicate ...)

wilzbach commented:
@MartinNowak I am too lazy to rebase all the time, so I removed the changelog entry again. Here is the message:

```
   $(LI $(XREF algorithm, sorting, multiSort) returns now a
       $(XREF range, SortedRange) instead of `void`)
```

MartinNowak commented:
> edit: btw I thought that if it's linked to an issue, you can generate the log entry automatically?

Yes, you'll get a Bugzilla fixed entry for references in a commit message. You can recognize that by the check mark in the dlang-bot comment.
https://github.com/dlang/phobos/pull/4235#issuecomment-214419022

MartinNowak commented:
> Have you looked at the implementation of predFun? It will break as soon as either lessFun(a,b) or lessFun(b, a) is true. Is there really a better way to do this?

Ah great, that's as good as it can get.

MartinNowak commented:
Auto-merge toggled on

CyberShadow commented:
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=16179
